src/fatx_disk.c

```diff
@@ -0,0 +1,294 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "fatx_internal.h"
+
+struct fatx_partition_map_entry const fatx_partition_map[] = {
+
+    /* Retail partitions */
+    { .letter = 'x',    .offset = 0x00080000, .size = 0x02ee00000 },
+    { .letter = 'y',    .offset = 0x2ee80000, .size = 0x02ee00000 },
+    { .letter = 'z',    .offset = 0x5dc80000, .size = 0x02ee00000 },
+    { .letter = 'c',    .offset = 0x8ca80000, .size = 0x01f400000 },
+    { .letter = 'e',    .offset = 0xabe80000, .size = 0x1312d6000 },
+
+    /* Extended (non-retail) partitions commonly used in homebrew */
+    { .letter = 'f',    .offset = 0x1dd156000, .size = -1 },
+
+};
+
+/*
+ * Given a drive letter, determine partition offset and size (in bytes).
+ */
+int fatx_drive_to_offset_size(char drive_letter, size_t *offset, size_t *size)
+{
+    struct fatx_partition_map_entry const *pi;
+
+    for (int i = 0; i < sizeof(fatx_partition_map); i++)
+    {
+        pi = &fatx_partition_map[i];
+
+        if (pi->letter == drive_letter)
+        {
+            *offset = pi->offset;
+            *size   = pi->size;
+            return FATX_STATUS_SUCCESS;
+        }
+    }
+
+    return FATX_STATUS_ERROR;
+}
+
+/*
+ * Return the disk size (in bytes).
+ */
+int fatx_disk_size(char const *path, size_t *size)
+{
+    FILE * device;
+    int retval;
+
+    device = fopen(path, "r");
+    if (!device)
+    {
+        fprintf(stderr, "failed to open %s for size query\n", path);
+        return FATX_STATUS_ERROR;
+    }
+
+    if (fseek(device, 0, SEEK_END))
+    {
+        fprintf(stderr, "failed to seek to end of disk\n");
+        retval = FATX_STATUS_ERROR;
+        goto cleanup;
+    }
+
+    *size = ftell(device);
+    retval = FATX_STATUS_SUCCESS;
+
+cleanup:
+    fclose(device);
+    return retval;
+}
+
+/*
+ * Return the remaining disk size (in bytes).
+ */
+int fatx_disk_size_remaining(char const *path, size_t offset, size_t *remaining_size)
+{
+    size_t disk_size;
+
+    if (fatx_disk_size(path, &disk_size))
+    {
+        return FATX_STATUS_ERROR;
+    }
+
+    if (offset > disk_size)
+    {
+        fprintf(stderr, "invalid disk offset\n");
+        return FATX_STATUS_ERROR;
+    }
+
+    *remaining_size = disk_size - offset;
+    return FATX_STATUS_SUCCESS;
+}
+
+/*
+ * Reformat a disk as FATX.
+ */
+int fatx_disk_format(struct fatx_fs *fs, char const *path, size_t sector_size, enum fatx_format format_type, size_t sectors_per_cluster)
+{
+    struct fatx_partition_map_entry const *pi;
+    size_t f_offset, f_size;
+
+    if (format_type == FATX_FORMAT_NONE)
```

Comment:
It's not immediately clear what format scheme of `NONE` means. It could be interpreted as blanking the partition, for instance. I think perhaps a better name would be `FATX_FORMAT_INVALID`? And in this case `fatx_disk_format` should return an error